### ISSUE TITLE
Blocking operations in the pipeline lower RabbitMQ performance dramatically.

### ISSUE THREAD

**Issue by timbussmann** [Particular/NServiceBus.RabbitMQ#330]
There is an edge case in the RabbitMQ transport which causes blocking code in the pipeline to block the transports message pump unnecessarily. This can occur in production e.g. when using NHibernate Persistence which does not offer fully async APIs. This can cause an endpoint to slow down to very low message throughput (e.g. a 50ms blocking call per message results in a throughput of ~20 msg/sec). Increasing the concurrency setting has no impact on throughput.

### How to Reproduce

Add any behavior or message handler which does a Thread.Sleep and no asynchronous operation.
e.g.:

    class BlockingBehavior : Behavior<ITransportReceiveContext>
    {
        public BlockingBehavior()
        {
        }

        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
        {
            Thread.Sleep(50);
            await next();
        }
    }

### Cause

The rabbitMQ client offers a event handler we subscribe to when a message to handle should be dispatched. The transports event handler is `async void` but there is no asynchronous operation happening before the blocking call. This causes the thread from the rabbitmq client which invokes the callback to block until an actual asynchronous operation is hit. Event when message handler use fully async code, this is too late in the pipeline, as the event invoking thread can be blocked by some persistence operations in the pipeline (e.g. the outbox). See the event handler implementation: https://github.com/Particular/NServiceBus.RabbitMQ/blob/develop/src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs#L155
(note that `semaphore.WaitAsync` will execute synchronously as long the semaphore limit isn't reached)

### Workaround

a `Task.Run` or `Task.Yield` as early as possible in the pipeline can be added using a custom behavior in the `ITransportReceiveContext` stage. This will free up the rabbitmq client's thread and allow further messages to be processed concurrently. This approach should also include increasing the configured concurrency limit (which works again with the workaround).

### Possible Fixes

a very simple approach could be to call `Task.Yield` as early as possible in the `Consumer_Received` handler to free up the calling thread. @bording raised a valid question whether this should be actually a core concern instead of a transport responsibility. Note that e.g. MSMQ uses Task.Run to invoke the pipeline for messages.

ping @Particular/rabbitmq-transport-maintainers @Particular/nservicebus-maintainers

**Comment by bording** [comment]
As I mentioned to @timbussmann when discussing this with him, this was something that had been nagging me about the current message pump implementation, but it never reached high enough importance for me to investigate whether it could actually be an issue or not. 😢

**Comment by nillis** [comment]
+1 I encountered this issue using NHibernate. I solved it in my case by adding the following behavior:
```
    class WrapsPipelineInAsyncTaskBehavior : Behavior<ITransportReceiveContext>
    {
        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
        {
            await Task.Run(async () => await next()).ConfigureAwait(false);
        }
    }
```

**Comment by danielmarbach** [comment]
@nillis 

try the following instead

```
 class WrapsPipelineInAsyncTaskBehavior : Behavior<ITransportReceiveContext>
    {
        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
        {
            await Task.Yield();
            await next().ConfigureAwait(false);
        }
    }
```

**Comment by nillis** [comment]
Both solutions results in the same throughput rate. Is there a reason to prefer the suggested solution?

**Comment by danielmarbach** [comment]
@nillis 

correct both achieve the same, it is probably more a preference thing. If you stick to the async / await usage then I think naturally my example is more aligned to that. If you want to remove the async declaration (state machine) then I'd go with this approach

```
    class WrapsPipelineInAsyncTaskBehavior : Behavior<ITransportReceiveContext>
    {
        public override Task Invoke(ITransportReceiveContext context, Func<Task> next)
        {
            return Task.Run(() => next());
        }
    }
```

**Comment by nillis** [comment]
Kk thx. Will this issues been resolved in this repo as well? As this issue occurs every time there is something blocking somewhere in the pipeline.

**Comment by danielmarbach** [comment]
@nillis Please see https://github.com/Particular/NServiceBus.RabbitMQ/pull/332

This resolves it for rabbit. Although there are two issues here, I would like to briefly discuss. 

1. The rabbit mq driver
1. The way async behaves

The rabbit mq driver

The rabbit mq driver uses a synchronous blocking model to push messages to us. Any blocking operation that is done on that path will automatically block the rabbit mq driver worker service by design. We can mitigate this by using the Task.Yield approach right before we process. This is a good compromise for people like you who are using some third party dependencies which are IO bound but of blocking nature. Unfortunately, this could impact truly asynchronous code paths since we are now explicitly yielding for each and every message received. In order to pull it completely in further discussions are needed with @Particular/rabbitmq-transport-maintainers 

Another option would be to fix it for example in NHibernate. So people that are using NHibernate would automatically get a behavior like outlined in https://github.com/Particular/NServiceBus.RabbitMQ/issues/330#issuecomment-276010830 at the beginning of the pipeline to offload. The caveat of this is that each plugin that uses this approach could potentially introduce more yielding which drastically deteriorates the throughput as well.

The way async behaves

By design async / await behaves so that anything executed before a completion port will be executed on thread entering the method (makes sense from an efficiency perspective). Only the caller or in some cases the callee has the context about what needs to be offloaded and what not. So in the case where you know you are calling something heavy (CPU bound or IO bound sync) in the handler, it is your responsibility to make the decision whether you want to offload to the worker thread pool and await the outcome sequentially or even do the heavy operations in parallel.

**Comment by adamralph** [comment]
@Particular/rabbitmq-transport-maintainers thoughts? Personally I think putting a `Task.Yield` is an acceptable price to pay to bridge from our async world to the sync world of the RabbitMQ.Client.

**Comment by adamralph** [comment]
FYI we are planning to release a 4.2.1 patch for this.

**Comment by nillis** [comment]
Great!

**Comment by adamralph** [comment]
FYI the proposed fix is in #335. Note the *War and Peace* comment explaining the logic. 💭

**Comment by adamralph** [comment]
FYI this [has been fixed in 4.2.1](https://github.com/Particular/NServiceBus.RabbitMQ/issues/340#issuecomment-278289141).

Thanks to @nillis @timbussmann and @danielmarbach for their efforts in helping to get this fixed. 🏆
